### ISSUE TITLE
Add an option to "Update Collection" generated from the APIs tab

### ISSUE THREAD

**Issue by tmdoit-zz** [a85/POSTMan-Chrome-Extension#6722]
**Is your feature request related to a problem? Please describe.**
When updating collections from API schema, all examples are lost - you need recreate all examples from zero - this kind of development is useless if you prefer develop collections from API spec.

**Describe the solution you'd like**
Add option "Update Collection" next to "Generate Collection" in APIs tab

**Describe alternatives you've considered**
After every update add manually all examples.

IMHO it's priority for APIs new feature you released.

**Comment by tmdoit-zz** [comment]
Sorry for confusion, now I see that response examples are generated from API spec.

**Comment by saivan** [comment]
But I still think this is probably valid, unless I missed something. What about keeping any tests we wrote in a collection?

**Comment by tmdoit-zz** [comment]
For me the point is to create API spec with examples and then based on it, generate collection  - API spec first approach. Currently one missing feature is to generate from spec more than one example if response has more than one possible returned value per status.
If I understand well, you are talking about collection first approach - generating API spec based on collection?

**Comment by saivan** [comment]
Well. I was going to use postman for my tests. So you generate the openapi file; then make a collection from it. After that, write tests in postman. It would be good to not have to lose all of your tests when you want to change the openapi file.

**Comment by preethammavin** [comment]
@saivan thanks for the feedback it is a valid use case, will keep it as a feature request

**Comment by tmdoit-zz** [comment]
@saivan now I get it - tests, not examples 🙂

**Comment by bikash8gupta** [comment]
The tests are deleted when a generate collection is applied. We definitely need a update collection because all collections have test suites in it and they are not kept when a new collection is generated.
Till then API beta is not of any use. Why I will keep maintaining API specification as well as collection from two places.

**Comment by hilary** [comment]
What's the status of this issue? I would also like to use Postman API spec first. Without an 'Update Collection' option, the 'Generate Collection' option is useless.

Generated specs better than nothing, but nowhere near as good as authored specs. I really don't see the point of 'Generate Collection' as it stands. Perhaps you could explain a maintainable workflow that used 'Generate Collection' and did not use 'Update Collection'?

**Comment by sankalp0o** [comment]
@hilary @bikash8gupta @saivan thanks for reaching out!

Collections and specifications (any of them) are very different from each other and at this point, we're unable to update a collection from the spec which can reliably work 100% of the time. That being said, it's not completely impossible depending on what has changed in the spec and how much has the collection evolved. There are solutions that can work in some of the cases. 

I request you to tell us the use cases that you have for updating the collection. What are you changing in the spec and how you imagine the collection to be updated?
E.g. I'm adding new endpoints to the spec and I expect new requests being created inside the collection following the correct folder structure.
E.g. I'm updating the descriptions for one of the endpoints and I expect the new description to replace the existing description in the collection request.

This will help us design a better solution! Please also let me know if you're open to getting on a quick call to understand your use case better.

**Comment by dvdvdmt** [comment]
I'm adding new endpoints or changing existing ones and I expect that different variable values such as `baseUrl`, `authToken` etc. will preserve in the collection settings.

**Comment by arlemi** [comment]
@sankalp0o Similar ask on our community forum: https://community.postman.com/t/automatically-sync-openapi-swagger-spec-to-collection/8551

Mostly, users are concerned about losing their tests and variables every time they regenerate the collection from the schema.

> The only concern I have with importing the swagger spec each time is that our test and variables will not be maintained, unless I’m missing something.
I believe even a 1 way sync that would update your Postman collection each time a change has been made to the swagger spec would be very useful.

I'll ask for more feedback if possible.

**Comment by andmoredev** [comment]
We use the Collection that gets generated by the API only for Documentation purposes so for my use case a full overwrite is completely acceptable. 
We keep our Test collections separately so that we can run tests for failures as well which would not be covered by an Open API definition file.

**Comment by hilary** [comment]
Even though I'm defining security in the spec, no auth is generated for the collection. That means I have to add auth to the collection after the fact. If generation were more complete, the lack of update wouldn't be as bad a problem.

We're primarily adding new endpoints, but occasionally deprecating existing ones. We're also updating documentation.

**Comment by tmdoit-zz** [comment]
Only difference between "Update Collection" and "Generate Collection" will be that "Update Collection" will recover tests written for defined collection endpoints. I use word "recover" because I don't see for this use case any reason to just not generate fresh collection from API spec and attach tests using collection endpoints as reference. I think that this way of solving problem will keep  implementation of this feature simple.

---
**UX:**
Use case: I want to update collection to keep written tests for collection endpoints
1. I click button "Update Collection"
2. In popup that appears I select via drop down list (API spec children) collection which I want to update
3. Postman updates collection preserving tests defined earlier for certain collection endpoints

Comments:
1. In popup windows information what collection update means
---

If I want to generate new, fresh collection without preserving tests I will use "Generate Collection"
If i want to generate new, fresh collection with endpoints tests preserved I will use "Update Collection"

Preserving variables could be solved same or similar way I think.

This should make sense for API spec first approach.

**Comment by tank104** [comment]
+1 for this - especially when prototyping - its cumbersome.  Right now I use Swagger, then when happy add it to Postman.  Would be nice to skip Swagger step.

**Comment by fabiendeborde** [comment]
> I'm adding new endpoints or changing existing ones and I expect that different variable values such as `baseUrl`, `authToken` etc. will preserve in the collection settings.

Pretty much this use case.
If I update the OpenAPI file (which is the origin of my collection), I don't mind if my collection is overwritten but I expect that my collections variables, mock server URL and tests are preserved.

**Comment by darrencapner** [comment]
From watching the recent webinars and reading documentation, there appears to be a workflow in mind, but I find that there is a gap that I have not been able to find a solution for, and would be solved potentially with this issue.

My use case is if I am wanting to iterate on an API design through design, generate collections, deploy, get feedback, and modify design....how would Postman be used if there is no way to update existing collections where test scripts were added for contract testing and performance testing:

One such workflow details:
1. Make an API definition
2. Generate collections for contract testing, performance testing, documentation, working collection, etc.
3. Get feedback to iterate on that API design as new endpoints are added, and changes to existing API contract is changed
4. Generate new collections for contract testing, performance testing, documentation, working collection, etc. keeping all test scripts in those new version collections.

Questions:
1. Do we have to copy tests from existing collections to newly generated collections?
2. Is there a way to generate collections over-top of existing collections, only modifying request contract as per the changes in the API spec?

**Comment by a85** [comment]
Quick update here: this workflow is top of our minds. 

We are looking at flows that allow you to:

1. Update a collection from an existing specification
2. Allow for pull request driven flows in case of conflicts between the new updated collection and the existing collection (so you can review and merge)
3. We are also looking at the reverse flow where you can update a collection and then optionally go backwards and update your specification (in the future, you will be able to do that with a pull request driven flow so you don't overwrite a schema right away)

Unfortunately, at the time being this is a manual process and honestly, we underestimated how popular this would be. Usage is exploding right now for the API tab and we are working hard to fix this loop. We just shipped an integration with GitHub and there is ongoing work with other repositories. 

I'd like to thank everyone for their patience and sharing use cases with us. It helps us define the scope quickly and get things to production.

**Comment by dpkirchner** [comment]
My use case is vastly simpler: I'm writing openapi 3.0 YAML and then I click Generate Collection so I can preview the result to ensure it is correct. I'm not even to the point that I'm writing tests. There is a documentation link in the "Develop" tab and it takes you to a page that looks correct on the surface, however it is missing the API responses. If that page included responses my use case would be satisfied.

**Comment by arlemi** [comment]
@dpkirchner To display example responses on the documentation you need to add examples in your collection. You can learn how to do that on our Learning Center: https://learning.postman.com/docs/postman/collections/examples/

**Comment by cdolan92** [comment]
There are a lot of use cases with Tests, but I'd like to focus on Documentation for our use case. We have a Published Collection running on api.ourdomain.com, and would like to keep that updated easily as our developers improve our API.

Here is our ideal workflow:

1. Commit to a Github repository linked with an API in Postman, using the new integration @a85 mentioned above, when there are updated endpoints.
2. Sync that API in Postman, which imported our OpenAPI 3.0 spec, to a Postman Collection.
3. Easily manage the Documentation associated with that Collection - If new endpoints are added, they would presumably have no Documentation/markdown associated with them. If endpoints change, the pre-existing Markdown would persist the update.

Basically, I'm not seeing how I can easily update our documentation when our team makes updates to our API, without having a bunch of markdown files kept in source control, and then copy+pasting those into each endpoint in the API in Postman... a very manual task I want to avoid!

**Comment by ghost** [comment]
I'm surprised this functionality doesn't exist. API specs are the heart and soul of rapid prototyping. Having to recreate docs, mocks, tests, and other artifacts when the spec changes are the antithesis of the desired workflow. Man, I hope this gets fixed. Kinda a dealbreaker for using Postman during the API design phase :(

**Comment by arlemi** [comment]
Glad you're bringing this up @sylvester-rubidium ! 😀 

We've started iterating on that feature and are looking for early feedback from users. Since this issue is full of users with good use-case for it we thought we'd ask here first. 
The session would be 45mins (over Zoom), starting with a short interview on how you use the API Builder and what you expect to achieve with it, followed by a usability test on a prototype.
If this is something you'd like to take part in, you can book time with us next week using [this Calendly link](https://calendly.com/arlemi/update-collection-ux-research) 📅, and if you have more questions please ask them here or directly to me on arlemi@postman.com. 

Looking forward to hearing from you all. 🙂

**Comment by arlemi** [comment]
Hey everyone, just a quick message to let you know that we're still looking for feedback from you all on the "update collection" flow. We've got time slots today, tomorrow, or Thursday and you can book any available time using this link: https://calendly.com/arlemi/update-collection-ux-research?month=2020-05. 😃

**Comment by fabiendeborde** [comment]
I would love to participate but I have a deadline this week, and so much work left to do....

If you still need some feedback in June I'll be glad to participate!

**Comment by arlemi** [comment]
@FabienDeborde No worries, thanks for your interest anyway! We may ask for a second round of feedback in June once we have a working version of the prototype, will update everyone here if so.

**Comment by hilary** [comment]
@arlemi @a85 Just booked our time! Wanted to say how happy we are that Postman has prioritized this workflow!

**Comment by gustave001** [comment]
+1 for this

**Comment by obieo** [comment]
I'd be happy to discuss what our expected workflow would be if there are dates opened up for June.

**Comment by arlemi** [comment]
Hey everyone! 

We've just released this feature in Postman 7.27, update your app (or [grab the latest version here](https://www.postman.com/downloads/)) and give it a try! You can find the full documentation in our Learning Center:
https://learning.postman.com/docs/postman/design-and-develop-apis/validating-elements-against-schema/ 

If you have any additional feedback please add it here or on[ the Postman community forum](https://community.postman.com/t/user-feedback-updating-api-elements/13308). We're already tracking a feature request for being able to trigger a validation/update through an API in #8697.

**Comment by gustavovendramini** [comment]
Thanks guys, excellent feature.

Testing here, and for me, the new one's methods/actions added in API doesn't appear to include on updating the collection

**Comment by arlemi** [comment]
@gustavovendramini Thanks for the feedback! We're tracking what you've mentioned and are hoping to release it in the next version of the app. 🙂

**Comment by molteninjabob** [comment]
@arlemi, @a85 et al., I just found this thread trying to find a solution for a different issue I'm experiencing. 

I've spent the last month building out an OpenAPI 3.0 schema for an existing api. The dev team already had a collection of requests (but no tests) and being able to validate the schema against an existing collection has been invaluable to ensure the schema I'm creating is correct. Great feature! I also just noticed the updated/improved visual diff, which is a nice addition to help troubleshooting.

I now find myself in an interesting position where I now have the schema created and I can generate a collection, but when I generate from the schema, all of the fields are set to the type that is expected instead of values that can be used, even if an example is provided. 

![Postman 2020-06-30 14-58-16](https://user-images.githubusercontent.com/11509961/86176508-4454ba00-bae2-11ea-9b10-865ebc53a62f.png)

![Postman 2020-06-30 12-22-43](https://user-images.githubusercontent.com/11509961/86176050-9fd27800-bae1-11ea-956e-c35cb61cf98d.png)

This might be a separate issue, though I think still related, but there doesn't seem to be a way to generate populated variables from the examples in the schema. The examples created in the collection are great for mocking responses, but if I want to exercise the requests directly in postman, I have to go replace all query parameters, path vars and request body values in every request. As you can imagine, this is just as time consuming as copying/pasting tests would be.

I'd love to see Postman take the example values defined in the API schema automatically added as collection variables and inserted into the generated requests instead of just putting the type. If done correctly, the generated requests would be immediately usable and look like this:

![Postman 2020-06-30 15-03-17](https://user-images.githubusercontent.com/11509961/86177018-289de380-bae3-11ea-9bb3-60924efbee5c.png)
![Postman 2020-06-30 15-04-45](https://user-images.githubusercontent.com/11509961/86177023-2b003d80-bae3-11ea-9627-eba0aec4d123.png)

**Comment by sankalp0o** [comment]
@molteninjabob thanks for bringing this up. We're putting in some configuration options so that you're able to select how you want your query params, request body etc to be generated. I'll post here what I have an update on this. If there's anything else that you'd like to see in the generate/update flows, do let us know!

**Comment by obieo** [comment]
Checking in on the "Update collection to retain tests" feature. I know this has been stated in multiple ways before but an ideal workflow for our team would be the following. 

1. OpenAPI spec is generated upon our CI build and our repo or CI pipeline in Azure DevOps would be integrated with Postman. 

2. Once integrated, the OpenAPI spec and Postman collection it's integrated with would be in sync each build to reflect changes in the spec. (New endpoints, deprecated endpoints, parameters, etc) 

3. Test and variables are retain each time the OpenAPI spec updates the collection in Postman

This workflow would allow for a seamless source of truth where our team could truly collaborate on test, development and documentations. 

Please let me know if something like this is being considered.

**Comment by arlemi** [comment]
@obieo You can already try out this feature, check out the documentation: https://learning.postman.com/docs/postman/design-and-develop-apis/validating-elements-against-schema/

Updating the generated collection would be a manual process for now but you'd still keep your tests etc. after every update. We're tracking a feature request to be able to do that flow using the Postman API in #8697.

**Comment by arlemi** [comment]
Closing this feature request as this has been out for some time now, we'll keep tracking improvements in other tickets to avoid confusion!

**Comment by obieo** [comment]
Hello – Thanks for the response however unless I’m missing something, this feature that is referenced does not hit on the scenario I was describing.

I see that you’re manually able to define the schema however what if I’m just looking to import the OpenAPI spec, using a url from the spec that I have already created a collection from.

When I go to import, I’m only left with 2 options (Copy, Replace). In this case I would not want to do either. I’d like to be able to import the same spec, pointing it to the existing collection generated from the spec, and update the schema automatically without deprecating the test that I’ve created for each test. I was over a url to some documentation but it does not cover being able to do what I just described.

Thanks,

Andrew Obaseki
Crowe LLP
Office: 317.706.2639 | Cell: 812.259.1829
andrew.obaseki@crowe.com<mailto:andrew.obaseki@crowe.com>
www.crowe.com<http://www.crowe.com>

**Comment by tank104** [comment]
I agree with Obieo,
I still don't see the feature to updte existing collection, unless I am mistaken.

**Comment by arlemi** [comment]
@obieo The current feature doesn't allow you to update a collection by importing a new version of the schema as a collection. It allows you to update collections generated from a schema in the API Builder. 
In your case you'd need to:
- Import your schema as an API
- [Generate a collection from it ](https://learning.postman.com/docs/designing-and-developing-your-api/the-api-workflow/#generating-a-collection)
- Copy over your tests from your current collection to the generated collection

Once this is done you'll be able to update your schema in the API Builder and update your collection while keeping the tests. Please [look at our documentation](https://learning.postman.com/docs/designing-and-developing-your-api/validating-elements-against-schema/#updating-api-elements) to see more details about that. 

FYI when you import a schema and create a collection from it all that we're checking is if there's a collection with a similar name, we're not checking the content of the collection, therefore we wouldn't be able to offer to update a collection this way.

**Comment by obieo** [comment]
Thanks @arlemi for the explanation. We will try this in the meantime. 

I do believe that a future enhancement of allowing users to update a collection based of changes to a schema, while maintaining the tests and variable would be very valuable. Especially as Postman continues to find was to automate the integration with OpenAPI. 

An example of this workflow would be 

1. Changes to API schema are made and committed 
2. OpenAPI spec is updated from Azure DevOps build 
3. Those changes would automatically sync to postman collection initially created in Postman (delta update) 
4. Tests, variables, pre-request, etc would remain intact for that collection

**Comment by arlemi** [comment]
@obieo From my understanding what's missing here is the ability to have the update done automatically or triggered through an API. We're already tracking two similar requests around the same: 
- Support OpenAPI-First, Forked Collection Workflow #8940
- Trigger an update to the collection from Postman API #8697

Give a 👍 on the original posts there if this is what you're looking for, this helps us prioritise! 🙂

**Comment by dvdvdmt** [comment]
@arlemi no it is not what @obieo described. Update of the Postman collection could be manual. 
In a simple form:
1. A user changes `openapi.yml`.
2. A user imports `openapi.yml` as a collection.
3. Existing collection updates but its tests, variables, pre-request remain intact.

**Comment by tank104** [comment]
@dvdvdmt is right - its just a manual button.  
So in our case where we are prototyping/change frequently the api spec, we can update what we have already, without everything being replaced.

**Comment by arlemi** [comment]
@dvdvdmt @tank104 Understood, and this is the whole point of this feature. I think there's a misunderstanding as to how you're supposed to use it. 

Postman allows API-first development and has a whole section of the app (called the API Builder) dedicated to authoring APIs as well as generating collection from a specific API schemas. You can create an API directly from the API tab or by selecting import as API instead of collection when importing a schema file. 
Once your API is imported, you can use the **Generate Collection** button to generate a collection, you can then either:
- link the API to a github repo to enable a 2-way sync between GitHub and Postman
- edit your schema directly in Postman
- copy/paste your schema file in Postman everytime there are changes

When this is done, you can check whether the collections previously generated need to be updated, and if so update them **without losing tests, pre-request scrips etc**.

You can see an explanation of the feature along with a demo on the API Builder webinar we ran a couple of months ago: https://youtu.be/Aolgf6tQaqQ?t=1378

**Comment by obieo** [comment]
@arlemi what you have posted makes sense however most of our teams aren't developing/authoring APIs within postman. 

- They are doing so in their own IDE and committing those changes to the repo which then updates the OpenAPI spec upon the build. 

- From there someone from the team (generally a tester) will go to Swagger, grab the spec url and import it into postman to generate the collection. 

- At this point it would be nice to recognize the url and the associated collection and allow the user to update the collection based of the change in the spec. 

Ultimately I agree with @dvdvdmt and @tank104 a good first step would just be a manual button to update the collection each time the spec is imported into Postman.

**Comment by cdolan92** [comment]
I agree with obieo - our team is pushing code out and we’re grabbing the Open API spec from the deployed code, which is then imported into Postman.

Once inside postman, updating the collection such that endpoint changes are accepted but documentation & markdown are preserved is the desired goal. This is particularly of interest because we’re using Postman’s hosted API documentation page.

The only work around to this problem is to create a new collection (C1) based off of updated OpenAPI docs, and clone my existing published Collection (C2, which is based off of C0).

Then, manually copy/paste endpoints from the newly generated Collection (C1) back into the copy of my original collection (C2), and publish  collection C2 as my primary collection for our hosted documentation page, and delete C0 and C1.

________________________________

**Comment by arlemi** [comment]
Ok, I think we're on the same page then! 

In both your use-cases, your Open API spec is pushed _somewhere_, if that somewhere is a GitHub repository you can then have it [automatically sync with Postman](https://learning.postman.com/docs/integrations/available-integrations/github/#syncing-your-api-schemas-on-github) (no need to grab it anymore) and all that is left to do is update the attached collections (i.e. "a good first step"). 

If you're using a different git repository then go give a thumbs up to the feature request #8245. 
If you believe we should have a way to detect when an import is made from a similar URL then please open a new feature request and we can keep the discussion going there. This one was specifically about adding a way to update collections from the API Builder which is now feasible. 🙂

**Comment by obieo** [comment]
Thanks @arlemi. I have created feature request #8950 for updating collections on manual imports.

**Comment by pankaj-cashify** [comment]
@arlemi I do not see any option on UI to update attached collections for OpenAPI. Am I missing something?

**Comment by arlemi** [comment]
@pankaj-cashify You should see a **Click to validate** button next to the collection you've previously generated, you need to navigate to either the Develop, Test, or Observe tab to see it. 

![updateelement](https://user-images.githubusercontent.com/5029719/92914543-a22b4980-f423-11ea-96ec-f21c70e9fcb2.gif)

**Comment by pankaj-cashify** [comment]
@arlemi sync is not working for me. I have attached a sample OpenAPI doc. I tried updating request parameter **app_version3** to **app_version5**

[test.json.txt](https://github.com/postmanlabs/postman-app-support/files/5208541/test.json.txt)

**Comment by abhinaba-ghosh** [comment]
Hi @pankaj-cashify we've been able to reproduce what you are facing. We have opened a bug for this issue. We will update this thread once we have more information. Thanks.

**Comment by abhinaba-ghosh** [comment]
Hi @pankaj-cashify , we have been able to reproduce what you are facing. We have opened a [new ticket](https://github.com/postmanlabs/postman-app-support/issues/9046) to track the issue. We will be posting updates once we have more information.

**Comment by gauravsoni1** [comment]
I see this issue has been closed, what would be the workaround for the following.
Suppose in v1 of Spec , i had defined 2 example. Later on I added another example to the same version. Is there some way to refresh the collection so that it would reflect the new spec ? Perhaps the right way would be a version update and a new collection ?

**Comment by arlemi** [comment]
@gauravsoni1 This is exactly what this feature allows you to do, once you have generated a collection you can keep on updating it as you update the schema. You can see an example of how to achieve that in https://github.com/postmanlabs/postman-app-support/issues/6722#issuecomment-691020293, and learn more about it in the Learning Center: https://learning.postman.com/docs/designing-and-developing-your-api/validating-elements-against-schema/#updating-api-elements

**Comment by molteninjabob** [comment]
> @molteninjabob thanks for bringing this up. We're putting in some configuration options so that you're able to select how you want your query params, request body etc to be generated. I'll post here what I have an update on this. If there's anything else that you'd like to see in the generate/update flows, do let us know!

@sankalp0o I see that this issue has been Closed. Regarding the configuration options you mentioned, do you have a reference to another ticket where this is captured? I'd hate for this to get lost because this issue was closed.

**Comment by arlemi** [comment]
@molteninjabob We've added more options to collection generations in Postman 7.30 which was released beginning of August. Here's some more info from the release notes: 

"You can [generate collections](https://learning.postman.com/docs/designing-and-developing-your-api/the-api-workflow/#generating-a-collection) from your API specification using the Postman API Builder. Though, we've listened to some of your feedback around being able to customize how these collections are generated, therefore from 7.30 when generating a collection you can choose:

- What should be use for the naming requests: `url` or fallback (`description`, `operationid`, or `url`)
- The indent character: spaces or tabs <small>(we all know tabs is better)</small>
- Whether to collapse the folder that have only one request or not
- Whether to optimize for conversion or not
- What should be used for `Request` and `Response` parameter generation
- If the folder organization should be using Paths or Tags

This configuration options are available when generating a collection from the schema, Define, Develop, Test, and Observe tabs."

**Comment by JerryChin** [comment]
I'm using postman for API documentation purpose, the API schema changes rapidly, but there's no way to update the linked collection.

**Comment by cdolan92** [comment]
@jerry Chin it seems to me that supporting a merge or other method with the Collection is not the standard use case for Postman. We tried making this work off and on for over a year, and are now moving away from postman to host our api docs because even though we generate OpenAPI compatible files, Postman performs some vendor lock in with a 1 way migration to Collections.

Once your schema (and particularly your markdown or other content) is in a collection, it’s very funky and hard to update it in any meaningful way other than to hand type it out again.
________________________________

**Comment by dengshenkk** [comment]
What is the status of this case ?

**Comment by arlemi** [comment]
@dengshenkk This is already released, check out this comment: https://github.com/postmanlabs/postman-app-support/issues/6722#issuecomment-647667330

@JerryChin What do you mean there's no way to update the linked collection? Are you not seeing the "Click to Validate" option? Giving a bit more details would help.

**Comment by Kazmirchuk** [comment]
is my understanding correct that this feature can be used only with OpenAPI v3? Unfortunately, I'm stuck with OpenAPI v2

**Comment by arlemi** [comment]
@Kazmirchuk Correct, we will likely add more over time but I don't have any insights into that at the moment.

**Comment by AlissonRS** [comment]
@arlemi I see the "Click to Validate" option, but I get "Unable to validate" because the schema is larger than 5 MB (it's indeed a large schema but it's actually around 2 MB, not 5). Is there a workaround for this?

Unfortunately I do not own the schema, otherwise I'd split it into smaller ones, but I really wanted to be able to Update this collection without losing variables/headers and other stuff in place in the existing collection.

![image](https://user-images.githubusercontent.com/4753901/142480861-6877dc6f-70e8-42c0-9b7e-736e28564f63.png)

**Comment by mklueh** [comment]
**If it is not possible to update a collection, what is the correct workflow then if an API changes?**

Let's say, you have multiple services and imported their specs into collections. You have created multiple tests,
and configured variables, and now the API changes. What to do now?

Is it like this:

1. Create a fork
2. Import the API a second time into a new collection
3. Copy everything from the new collection into the old collection
5. Delete new collection
6. Merge back

What about tests / variables for example?

**Comment by arlemi** [comment]
@AlissonRS Can you please open a new ticket to raise that issue? If the schema is smaller than 5MB you should be able to update the collection. If you're able to provide the schema it would also be super helpful! 

@mklueh If you have the API definitions, why are you not using them to generate the collections, and then update them as the API definitions evolve? I'm not sure if I understand your use-case here.

**Comment by mklueh** [comment]
> @AlissonRS Can you please open a new ticket to raise that issue? If the schema is smaller than 5MB you should be able to update the collection. If you're able to provide the schema it would also be super helpful! 
> 
> @mklueh If you have the API definitions, why are you not using them to generate the collections, and then update them as the API definitions evolve? I'm not sure if I understand your use-case here. 

@arlemi I'd love to do that, but I don't see any option to update an existing collection based on a new OpenAPI definition in Postman. Maybe I'm just overseeing it?
I was asking for a kind of workaround based on my assumption, that there isn't such an option.

**Comment by arlemi** [comment]
@mklueh As long as the collection has originally been generated from that spec (and that they are linked) you should be able to do it. Have a look at the documentation here: https://learning.postman.com/docs/designing-and-developing-your-api/validating-elements-against-schema/#validating-elements

**Comment by mklueh** [comment]
@arlemi thanks, this brought me one step further towards my goal. 
Not sure what you mean with "linked", but I've imported it with the import assistant by pasting my openapi spec URL. When I press validate, it does not pick up the latest spec the API provides through its spec URL, but rather the old one I've used during the import. 

The validation works fine if I copy and paste the new definition into the editor of the definition tab, but this can't be the right way I guess.

**Comment by arlemi** [comment]
@mklueh Glad you got it working! And it is the expected flow, when you import an API definition it's a one time import and it doesn't keep a reference to the original link. If you'd like to have your API definition synced with your Git repo (Github/Gitlab/Bitbucket) have a look at this: https://learning.postman.com/docs/designing-and-developing-your-api/versioning-an-api/#using-an-external-git-repository

**Comment by cavator** [comment]
so a big company like postman, could not create a feature to "watch" or update the API? in 2 years? that could not be serious.
or can and I'm missing something? and i don't  want to connect to some repo, i want it to watch my local file that i set when create the API
it's not that hard create this, even more in 2 years of issue

**Comment by cavator** [comment]
someone could please open this issue again, it's not even close to be closed.

**Comment by arlemi** [comment]
@cavator This is about updating a collection that was generated from an API in Postman. Looks like you're looking for something else so I'd recommend creating a new issue and provide as many details as possible.

**Comment by cavator** [comment]
@arlemi 
it's not because because i said "watch", that you will pretend i'm talking about something else, "or update the API" i think we talking about the same thing, update the collection generated from an API in postman, people now are 2y updating 1 line of the mutation or query and deleting and recreating the API. so please re-open this issue, like i said, it's not even close to be done. and i can't belive a big inc like postman is taking 2y to add 1 button, and still acting like its all ok

**Comment by mklueh** [comment]
I agree. The workflow would be much nicer if we could tell postman to re-fetch the new schema and get a git-like diff between the previous and new schema provided, to decide if we want to accept the new schema entirely or merge. 
Then Postman could automatically create a new API version based on the version provided in the schema.

It's not that the current workflow won't work, but the auto-updating feature would be more intuitive and less of a hassle if your API gets frequently updated. 
This is honestly something I was expecting to happen when I used Postman for the first time until I found out that manual work has to be done.

**Comment by arlemi** [comment]
@cavator Check that video from 5:40 as it shows the update flow in Postman: https://youtu.be/BUZiRtGRHj0?t=340

If you're not seeing this you might want to check if your version of Postman is up to date (latest is 9.7.0).

@mklueh Sounds like you're interested in auto-updating a schema in Postman from a schema in Git (Github/Gitlab...) rather than between a schema and a collection. If that's the case I'd recommend[ watching this video](https://www.youtube.com/watch?v=rTTAVlUxfDc) and/or [reading this blog ](https://blog.postman.com/the-reimagined-api-first-workflow-for-developers/) as they show how you can integrate with GitLab/GitHub/Azure DevOps and get diffs of changes between your repo and your API in Postman. 
If you have more questions on this I'd recommend chiming in on the community forum instead as you'll get more eyes than on a closed issue here. 😅 https://community.getpostman.com/

**Comment by tbcarver** [comment]
Just to validate your workflow for API schema first development.
1. Create an API Schema and make some changes.
2. Click Generate Collection button (well placed on the upper right above the tabs on the api schema page). Now I have a linked collection.
3. Make edits to the API Schema.
4. Ignore Generate Collection (pretend it doesn't exist), go to the documentation tab, look to the right for validate schema button, click button and wait for failure, click Review Issues button, on the sync collection to schema page, check select all and then hit the Confirm Changes to the Collections button.

Wow you nailed this common workflow, not overly engineered or confusing at all.  It was real obvious after digging through docs, this post, videos etc.  Have you considered hiring a UX expert?  Also, are you recommending not using the API Schema feature because clearly you have wedged this feature into collections like some type of ugly side car.

Sorry about the sarcasm, but I think you should consider real world workflows before releasing your product. BTW people do updates to the API Schema pretty often, just fyi.

**Comment by mklueh** [comment]
> @cavator Check that video from 5:40 as it shows the update flow in Postman: https://youtu.be/BUZiRtGRHj0?t=340
> 
> If you're not seeing this you might want to check if your version of Postman is up to date (latest is 9.7.0).
> 
> @mklueh Sounds like you're interested in auto-updating a schema in Postman from a schema in Git (Github/Gitlab...) rather than between a schema and a collection. If that's the case I'd recommend[ watching this video](https://www.youtube.com/watch?v=rTTAVlUxfDc) and/or [reading this blog ](https://blog.postman.com/the-reimagined-api-first-workflow-for-developers/) as they show how you can integrate with GitLab/GitHub/Azure DevOps and get diffs of changes between your repo and your API in Postman. If you have more questions on this I'd recommend chiming in on the community forum instead as you'll get more eyes than on a closed issue here. 😅 https://community.getpostman.com/

No, that is not what I meant. I meant "git-like diff" for the changes coming from an OpenAPI endpoint

The way I'd like it to be:

> 1. Make your API changes and raise the schema version
> 2. Tell postman to reload changed schema, auto-create a new API version based on the given schema version, then update the collection.

**Comment by eladchen** [comment]
Correct me If I'm wrong, but updating the example of a property is not detected when validating the documentation...?

I'm using the v9.13.0.

Is there any docs on what type of changes are detected when "validating" a documentation/collection?

**Comment by arverma77** [comment]
@arlemi sync is not working. I see that when we update the API name , only this particular change is taken into consideration. But if I try updating request parameter, the sync doesnt work. A ticked was logged for this issue in year 2020.

https://community.postman.com/t/user-feedback-updating-api-elements/13308

Can you please let me know how to sync a change to an attribute in API file to the collection.

**Comment by arverma77** [comment]
> Correct me If I'm wrong, but updating the example of a property is not detected when validating the documentation...?
> 
> I'm using the v9.13.0.
> 
> Is there any docs on what type of changes are detected when "validating" a documentation/collection?

Only the attributes which has property required=true is validate for any change. Optional attributes are currently not validated against the API changes.

**Comment by liviusighiartau8x8** [comment]
Is there a way to sync Postman collection with OpenAPI/Swagger? I think this was raised in the past, has something been done for it?

**Comment by n4j1Br4ch1D** [comment]
No Solution Yet?

**Comment by arlemi** [comment]
Hey everyone,

At the moment it's not possible to update collections that have been generated from an OpenAPI definition in Postman v10 through the UI. We're working on a newer version of the feature that should fit better in the flow of an API producer. 

While we re-implement it, you can still use the Postman API to update your collections, see more info here: https://www.postman.com/postman/workspace/postman-public-workspace/request/12959542-b5acf70c-7635-4d64-b17b-d4b2824c4a39?ctx=documentation

I'm re-opening this issue and we'll keep posting updates here as we make progress. cc @shashankawasthi88

**Comment by AmirHmZz** [comment]
Any updates on this?

**Comment by shashankawasthi88** [comment]
Hey @AmirHmZz we are actively working on this issue and will update the thread as soon as we have an update for you folks.
Thanks.

**Comment by Bandgren** [comment]
@shashankawasthi88 is this weeks, months, half a year away? Would be good to have a rough estimate so we can either put time into use the API endpoint or just wait for it to be implemented back! :)

**Comment by shashankawasthi88** [comment]
@Bandgren this is in our short-term road map. I can not comment on the exact timelines yet, but we are actively working on the feature request.

**Comment by sc-atompower** [comment]
With the integrations to Azure API Management, etc, we are actively looking forward to this.

**Comment by Matvii-Boichenko** [comment]
On project we are using AWS APIGateway & Postman as documentation, feature with "git-like updating" of collection from imported OpenAPI Definition of Postman API would be VERY Handy

**Comment by Bandgren** [comment]
@shashankawasthi88 we are now almost 2 months from my original query. This is a "check-in" how is this feature going? Any progress you can update us on?

**Comment by shashankawasthi88** [comment]
@Bandgren and all, this feature is being released in our upcoming March 15th release. We look forward to you using and providing your valuable feedback on the same.

**Comment by Bandgren** [comment]
@shashankawasthi88 Thank you so much for the update, and that is some really great news!

**Comment by Bandgren** [comment]
@shashankawasthi88 was this released with this release? Cant find anything in the release notes, have tried to update a openapi spec in both the mac app and in chrome and cannot find any "update collection" button anywhere?

**Comment by fmiqbal** [comment]
@Bandgren yes it is released,
![image](https://user-images.githubusercontent.com/8409391/225606018-a1039c07-21a1-4a0e-b27c-48003fc40fd5.png)

the update button is a bit hidden though, it only visible when there's schema change, and it run in background, so wait a bit after you save your change in your schema, and this should pop out to the side of your collection 

![image](https://user-images.githubusercontent.com/8409391/225608599-db7c09f5-0d8a-4236-ae55-9b4ebf06ba1e.png)

**Comment by arlemi** [comment]
@fmiqbal Good catch! It has been released in the last version and is currently being rolled out to users over the next few days. 

You can read more about how to update the collections here: https://learning.postman.com/docs/designing-and-developing-your-api/developing-an-api/adding-api-elements/#keeping-a-collection-in-sync-with-an-api

And we'll soon be releasing a video demonstrating how to do it too, I'll post it here when it's out!

**Comment by Bandgren** [comment]
@fmiqbal & @arlemi thank you so much! Looking forward to start using it!

**Comment by Alchoder** [comment]
From what I understood, automatic synchronization is only available when you have Enterprise workspace ?

**Comment by shashankawasthi88** [comment]
@Alchoder this is available for all.

**Comment by AmirHmZz** [comment]
@shashankawasthi88 I don't use postman enterprise and still I cannot figure it out. Can someone make an step to step guide to make it work?

**Comment by arlemi** [comment]
@AmirHmZz This video shows you a step-by-step of updating a collection generated from an API definition: https://www.youtube.com/watch?v=pbkS4XVsUNI

Please make sure you're on the latest version of Postman (v10.12), if not you can [download it from here](https://postman.com/downloads/) or use [Postman in your browser here](https://go.postman.co).

**Comment by wooseongshin** [comment]
@arlemi 
Is this not available in the CLI yet?

**Comment by arlemi** [comment]
@wooseongshin Having this in the CLI would be a different feature request, I'd recommend opening your own explaining your use-case and what you'd expect the CLI to be doing. 

I'll close this issue now that the collection update is available to everyone. 👍

**Comment by Andreas-T** [comment]
Adding or removing paths does work and triggers a sync notification, as per the video, but changes within an existing path does not. Neither changes in the description, content or response examples are detected.

EDIT: Upon further testing it seems that value changes for existing properties in response examples are detected but new properties are not, even if the schema is updated to reflect the changes.

**Comment by sujayvenaik** [comment]
but changes within an existing path does not. Neither changes in the description, content or response examples are detected
@Andreas-T can you give one example for this? When you say within an existing path does not: my first hunch is the routes might have wildcards and might be getting matched to some other route. But ill wait for you example to confirm. 

-- 
Upon further testing it seems that value changes for existing properties in response examples are detected but new properties are not, even if the schema is updated to reflect the changes

Can you provide an example if possible?

**Comment by johnny32** [comment]
Is this not supported for OpenApi 3.1 definitions?

**Comment by GRustamyan-ST** [comment]
> @wooseongshin Having this in the CLI would be a different feature request, I'd recommend opening your own explaining your use-case and what you'd expect the CLI to be doing.
> 
> I'll close this issue now that the collection update is available to everyone. 👍

Does this work for JSON definitions as well?
I tried same API yaml and JSON definition - yaml works, JSON was not working.

**Comment by shashankawasthi88** [comment]
@GRustamyan-ST it works both for JSON and YAML. Can you share a screen recording of the scenario that is not working for you?

**Comment by beckmarc** [comment]
I imported my collection via a URL that contains a valid Open API 3 spec in json and yaml format (I tried both). The importing works but when I click on the API to add a collection the "Generate from definition" button is greyed out. Any idea why this happens? I want to sync my API spec with Postman so my frontend devs can easily interact with the API.

**Comment by shashankawasthi88** [comment]
@beckmarc - can you check for any syntax issues in the spec file that is imported?

**Comment by beckmarc01** [comment]
@shashankawasthi88 - Yes, I already did using this editor [editor.swagger.io](https://editor.swagger.io/). However, in response to your comment I tried out a new one from swagger [editor-next.swagger.io](https://editor-next.swagger.io/) and it showed me some new errors. I fixed them and now it works.

**Comment by shashankawasthi88** [comment]
@beckmarc the Postman schema editor can point out the syntax error as well. Learn more [here](https://learning.postman.com/docs/designing-and-developing-your-api/developing-an-api/validating-elements-against-schema/#validating-api-definitions).

**Comment by vladiangelov** [comment]
@shashankawasthi88 great work - are there any plans to extend this functionality to GraphQL APIs?

**Comment by shashankawasthi88** [comment]
@vladiangelov yes we would extend it to GraphQL as well. Do you mind creating a new ticket for it? This one has a lot of different conversations and reopening this will be misleading.

**Comment by vladiangelov** [comment]
Thanks @shashankawasthi88 - I've followed your advice and created a new issue (as you can see above).

**Comment by reloxx13** [comment]
this was free in v9 by revalidating the schema and collection, then it was disabled to reimplement it for money milking.

**Comment by golkedj** [comment]
This was closed but does not add the ability to do this for a collection created from swagger documentation, which seems to be the most requested feature here. Is there a plan for this going foward?

**Comment by XedinUnknown** [comment]
Yep, I'm doin it. Any progress or plans for this?

**Comment by soumyaDarshanee** [comment]
Is there a plan to implement updating a postman collection created from Swagger , yet ?

**Comment by arlemi** [comment]
This is already possible, please follow the instructions in the Learning Center here: https://learning.postman.com/docs/designing-and-developing-your-api/developing-an-api/adding-api-elements/#keeping-a-collection-in-sync-with-an-api

**Comment by sergeiakopov-wellsky** [comment]
> This is already possible, please follow the instructions in the Learning Center here: https://learning.postman.com/docs/designing-and-developing-your-api/developing-an-api/adding-api-elements/#keeping-a-collection-in-sync-with-an-api

> On the API's overview, select the more actions icon More actions icon next to a collection and select Enable update suggestions from definition.

There is no such thing as `Enable update suggestions from definition` anywhere on the API overview page or any menu.

**Comment by solo812** [comment]
Is the ability to sync a collection available for any other tier than enterprise? Im a solo developer and am looking for this for my personal projects.

**Comment by arlemi** [comment]
@solo812 This is available for all plans, on the free tier you have the ability to create up to 3 APIs which should be enough for personal projects. 🙂
